pageserver/src/page_service.rs

```diff
@@ -482,7 +482,21 @@ impl PageServerHandler {
         let _enter = info_span!("get_page", rel = %req.rel, blkno = &req.blkno, req_lsn = %req.lsn)
             .entered();
         let tag = RelishTag::Relation(req.rel);
+        let latest_gc_cutoff_lsn = timeline.get_latest_gc_cutoff_lsn();
```

Comment:
I added latest_gc_cutoff_lsn: &RwLockReadGuard to `wait_or_get_last_lsn` and `check_lsn_is_in_scope` methods,
and perform comparison with gc_cuttof here.  Returning guard by `wait_or_get_last_lsn`  seems to be more problematic (I am not sure if it ever possible) and even if it is possible I think that explicit specification of latest_gc_cutoff_lsn lock scope is more straightforward and clear solution.

I do no think that delaying start of GC by `handle_basebackup_request` is really problematic. Size of backup is small (not  depends on size of the database) and postponing GC for some time is not critical.

Reply:
> I added latest_gc_cutoff_lsn: &RwLockReadGuard to `wait_or_get_last_lsn` and `check_lsn_is_in_scope` methods, and perform comparison with gc_cuttof here.

Ok.

> Returning guard by `wait_or_get_last_lsn` seems to be more problematic (I am not sure if it ever possible) and even if it is possible I think that explicit specification of latest_gc_cutoff_lsn lock scope is more straightforward and clear solution.

It'd be pretty straightforward to return the guard. But I'm ok with this, too.

> I do no think that delaying start of GC by `handle_basebackup_request` is really problematic. Size of backup is small (not depends on size of the database) and postponing GC for some time is not critical.

Agreed, this is good enough for now. WIth parking_lot RwLocks, which offer Fairness, this could be a problem though:

1. The primary compute node is very active, and issues a lot of GetPage@LSN requests
2. A new read-only node is created, which calls 'basebackup'
3. The basebackup acquires the read-lock on `latest_gc_cutoff_lsn`
4. GC starts, and tries to acquire write-lock on `latest_gc_cutoff_lsn`. It blocks because basebackup holds the lock
5. All new GetPage@LSN requests are blocked behind GC trying to acquire the write-lock, until basebackup finishes.

That could be a pretty noticeable "hickup".

Currently, we use std::sync::RwLock, which doesn't guarantee fairness so should not suffer from that problem. However, you could get starvation instead: if there's a constant stream of GetPage@LSN requests, GC might never get a chance to run.

I think that's OK for now, but let's add a comment somewhere explaining how this lock is used, and describe these potential issues. In the long run, it would be nice to have a more advanced RWLock-like mechanism, where GC could wait for all the already-running GetPage requests to finish, but not for new ones.

Reply:
I have to agree with you: such priority inversion can really take place in this case.
I wonder if `parking_lot` implementation is able to take in account thread priorities and try to address priority inversion problem. I suspect that answer will be no.